src/sentry/search/snuba/backend.py

```diff
@@ -343,19 +375,22 @@ def do_search(project_id, environment_id, tags, start, end,
         filter_keys=filters,
         aggregations=aggregations,
         orderby='-' + sort,
-        limit=limit,
         referrer='search',
     )
 
     # {hash -> group_id, ...}
-    hash_to_group = dict(
-        GroupHash.objects.filter(
-            project_id=project_id,
-            hash__in=snuba_results.keys()
-        ).values_list(
-            'hash', 'group_id'
+    if candidate_hashes is not None:
+        # any hash coming back had to come from our candidate set
+        hash_to_group = candidate_hashes
+    else:
+        hash_to_group = dict(
```

Comment:
Similar to your comment on #8675, this could cause a `KeyError` on line 402 if this is an "old" non-hash "hash" that was upgraded — it'll be in Snuba (because it was upgraded during backfill) but not in PostgreSQL (since it wasn't upgraded during write), so this can return a result set that's smaller than the input set. Probably worth protecting that key access, logging it as a warning and continuing the loop?

Reply:
Good catch.

This is part of a bigger issue where I'm not handling `GroupHashTombstone`s at all in search, which would (currently) require that I use the `issues` API. Have to do some thunking here because I think my fix will also (necessarily) take care of this.

Reply:
I still need to fix the tombstone thing, but just to revisit this so I don't forget: we already check whether the key is in the dict and log in the else branch, so I don't think this could blow up?

https://github.com/getsentry/sentry/blob/950ced3ca7b13059417753a97274eb924f654432/src/sentry/search/snuba/backend.py#L399-L400

Reply:
😅 Yup you're right about that one too. Sorry!